### ISSUE TITLE
Add a way to capture host error/trace output

### ISSUE THREAD

**Issue by pakrym** [dotnet/core-setup#4455]
Currently, when hosting coreclr in AspNet Core Module we have to try and setup in-process stdout/error redirection to be able to capture startup error (framework not found, etc).

It's not easy to do correctly and is not very reliable. Would be nice to have better support for capturing/forwarding this information.

/cc @jeffschwMSFT

**Comment by jeffschwMSFT** [comment]
cc @vitek-karas

**Comment by vitek-karas** [comment]
@pakrym - would having an option to write it to a file instead work for you?
I'm not sure I can think of any other way to output the tracing from the native process (other than inventing some kind of native plugin mechanism).
Do you think having 2 options:
* stderr (or possibly stdout)
* file
Would be enough?

**Comment by pakrym** [comment]
We load hostfxr into the w3wp process and call native functions (hostfxr_main) directly so stderr/out is pretty hard to redirect.

File might be an OK workaround.

Another solution is for hostfxr_ methods to take a buffer and write output to it

**Comment by vitek-karas** [comment]
I'm not sure we can "Extend" the existing hostfxr methods. So I'm thinking more like a registration API - hostfxr_register_trace_listener(IHostTraceListener). I assume that would work for your scenario right?
There's also the consideration that today the tracing is across all the hosts (dotnet/apphost, hostfxr, hostpolicy) so this would have to propagate.

**Comment by pakrym** [comment]
Absolutely, `hostfxr_register_trace_listener` would be perfect for us.

**Comment by jkotas** [comment]
Would it make sense to send this to ETW trace instead of defining a custom callback? ETW trace is more general - anybody can subscribe to it, even the folks who are not in charge of the hosting.

**Comment by pakrym** [comment]
As long as there is an easy way of extracting useful information to display to the customer about why their app wasn't able to start.

**Comment by jkotas** [comment]
What is the pipeline that you are using to display useful information to the customer? https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/azure-diagnostics ?

**Comment by pakrym** [comment]
We write an event to the event log and forward standard output to stdout log file.

**Comment by jkotas** [comment]
How is it gong to work end-to-end? Let's say I have deployed an app with IIS Module to Azure, it does not start and I suspect that the problem is in the CoreCLR hosting. What do I need to do to turn on this logging and how do I access the log?

**Comment by pakrym** [comment]
In the case of AppServices 2 ways currently:

1. Enable `stdoutLogFile` in `web.config` (https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/aspnet-core-module?view=aspnetcore-2.1) and use SCM (Advanced Tools in the portal) to look at it when the site fails to start.
2. Use portal to access event logs:
![image](https://user-images.githubusercontent.com/1697911/45655002-eaad0800-ba92-11e8-9250-9d305c8da6c4.png)

**Comment by jkotas** [comment]
How do you expect this new logging to be enabled for Azure? I suspect that it is way too verbose to be enabled by default.

**Comment by pakrym** [comment]
We always log app startup errors.

**Comment by pakrym** [comment]
We would also love to log unhandled exceptions. `Program.Main` exceptions are not going to event log because we catch them in ACNM to prevent worker process from crashing.

**Comment by jkotas** [comment]
Make sense.

I would scope this issue to just providing details and overriding behavior for *fatal errors* (ie. errors that are printed to stderr by default today). But I think we should extend the design to cover all fatal errors, not just fatal errors coming from core-setup part.

Thoughts?

**Comment by pakrym** [comment]
Sounds reasonable. Fatal errors are what we mostly care about.

**Comment by jkotas** [comment]
> Does for example unhandled managed exception count?

Yes.

**Comment by vitek-karas** [comment]
I know about that problem, but that one is bigger as it also tears down the process (we call ExitProcess on unhandled managed exception), which the ASP.NET needs to avoid.

That said if we would still send the fully detailed error info to the new channel, it might partially solve that issues as well.

**Comment by vitek-karas** [comment]
Looked at how CLR does reporting of unhandled exceptions and it basically just writes to STDERR. We could add the ability to "forward" this output to the host somehow.
As far as I can tell this would potentially redirect all STDERR from the runtime - not just unhandled exceptions. We could basically highjack the [PrintToStdErr](https://github.com/dotnet/coreclr/blob/2aa1b3c4c02ccff960fef72fe96175d9f62bb6d9/src/vm/util.cpp#L456) (and friends).

The downside is that this would be thread agnostic - so the host would get all messages from all threads with no ordering guarantees (just like we write it to console).
This would be unlike our potential solution for the hostfxr/hostpolicy which will be thread aware (it has to be as it's going to be used also for the runtime calling into host).

**Comment by vitek-karas** [comment]
Regarding the host error reporting (not the runtime), we ran into the need for redirecting errors as well. So I played with some possible solutions and this one seems like the best option:

`hostpolicy` would expose this new export
``` C++
typedef void(*corehost_error_writer_fn)(const pal::char_t* message);
void corehost_set_error_writer(corehost_error_writer_fn error_writer);
```

`hostfxr` would expose basically the same thing, just different name
``` C++
typedef void(*hostfxr_error_writer_fn)(const pal::char_t* message);
void hostfxr_set_error_writer(corehost_error_writer_fn error_writer);
```

The function would register the specified callback on the current thread. The registration would be per thread (thread local), this is needed for it to work correctly when used with the other exports (not the main "run this" exports). Any error which would normally be written to `stderr` would now only call the callback with the message (no output to `stderr`).

For the calls to "run this" exports, all the error reporting happens on the main thread anyway (before we call into the runtime). The fact that the managed code can than have multiple threads will not cause issues, since there's currently no codepath in which the other threads would report errors from the host libraries.

Only one callback can be registered, the method simply overwrites the one set before. Setting the callback to `null` unregisters the callback and any future errors will again go to `stderr`.

Registering the callback for `hostfxr` would also propagate this setting into any calls into `hostpolicy` the `hostfxr` may invoke. So it will be enough to set it on `hostfxr` if only `hostfxr` exports are used.

This has no effect on tracing (if tracing is enabled, errors are written to the trace output as well).

This feature is needed for the proposed `ComponentDependencyResolver` for `hostpolicy` exports, so something like this will have to be implemented, regardless of ASP.NET requirements. But it would be really nice if this would work for ASP.NET as well.

**Comment by pakrym** [comment]
This looks like it would work for us.

**Comment by sdmaclea** [comment]
@vitek-karas Looks like you merged #4816.  Is that sufficient to close this?  Or do we need something else?

**Comment by pakrym** [comment]
Would be nice to have some way of capturing unhandled exceptions too. @vitek-karas @jkotas  do you want me to file that separately?

**Comment by vitek-karas** [comment]
Since #4836 covers the runtime errors, this has been effectively resolved via #4816.
